File: ChainOfResponsibility/ChainOfResponsibility.py
from abc import ABC, abstractmethod

class Handler(ABC):
    def __init__(self, successor=None):
        self._successor = successor

    @abstractmethod
    def handle_request(self):
        pass

class ConcreteHandler1(Handler):
    def handleRequest(self):
        if True:  # Can handle request.
            print("Finally handled by ConcreteHandler1")
        elif self._successor is not None:
            print("Message passed to next in chain by ConcreteHandler1")
            self._successor.handleRequest()

class ConcreteHandler2(Handler):
    def handleRequest(self):
        if False:  # Can't handle request.
            print("Finally handled by ConcreteHandler2")
        elif self._successor is not None:
            print("Message passed to next in chain by ConcreteHandler2")
            self._successor.handleRequest()

class ConcreteHandler3(Handler):
    def handle_request(self, request):
        if request == 'request3':
            print('ConcreteHandler3 handles the request3.')
        elif self._successor:
            self._successor.handle_request(request)

# Client Code
ch1 = ConcreteHandler1()
ch2 = ConcreteHandler2(ch1)
ch2.handleRequest()

"""
Output:
    Message passed to next in chain by ConcreteHandler2
    Finally handled by ConcreteHandler1

"""

File: ChainOfResponsibility/ChainOfResponsibility.java
abstract class Handler {
    protected Handler successor = null;

    public Handler(Handler successor) {
        this.successor = successor;
    }

    public abstract void handleRequest();
}

class ConcreteHandler1 extends Handler {
    public ConcreteHandler1(Handler successor) {
        super(successor);
    }

    @Override
    public void handleRequest() {
        if (true) {  // Can handle request.
            System.out.println("Finally handled by ConcreteHandler1");
        } else if (successor != null) {
            System.out.println("Message passed to next in chain by ConcreteHandler1");
            successor.handleRequest();
        }
    }
}

class ConcreteHandler2 extends Handler {
    public ConcreteHandler2(Handler successor) {
        super(successor);
    }

    @Override
    public void handleRequest() {
        if (false) {  // Can't handle request.
            System.out.println("Finally handled by ConcreteHandler2");
        } else if (successor != null) {
            System.out.println("Message passed to next in chain by ConcreteHandler2");
            successor.handleRequest();
        }
    }
}

public class ChainOfResponsibility {
    public static void main(String[] args) {
        ConcreteHandler1 ch1 = new ConcreteHandler1(null);
        ConcreteHandler2 ch2 = new ConcreteHandler2(ch1);
        ch2.handleRequest();
    }
}
